rfcs/006-package-management.md

```diff
@@ -0,0 +1,359 @@
+follow this, pre-1.0 versions would never get binned together. `cargo` modifies
+the semver rules, allowing `0.x.y` versions to be binned together if they share
+the same `x`. Should we do the same?
+
+## The registry
+
+How should we manage the global registry? There's a potential for incurring
+substantial maintenance costs here, so we should be careful.
+
+We will provide a git repo, hard-coded to live at `github.com/tweag/nickel-mine`,
+to serve as the registry.
+This repo will contain the "index", but not the actual package contents. It
+will contain one file per package, each of which contains a line per version.
+Each entry specifies the location of the package (currently required to be on
+github) and its git tree hash. This ensures that packages are immutable, but it
+doesn't stop them from disappearing: we don't keep a copy of the actual package
+contents.
+
+The registry entries are named like "github/\<org\>/\<package\>" (where in the
+future we might support places other than github). This allows us to skip
+registration and authentication: if someone has the github permissions to create
+`jneem/foo` on github, they also have permissions to create and update the
+`github/jneem/foo` nickel package.
+
+We will provide a backend service to update packages. Users can submit a request
+to the backend, asking to publish version `0.6.5` of package `jneem/foo` at a
+specific git revision. The service will check that the github repo `jneem/foo`
+has a `v0.6.5` tag pointing at that git revision; if so, it will add it to
+the index.
+
+### Alternative: manual pull requests
+
+If package volume is low enough (which it probably is, at first), index updates
+could be done manually via pull requests.
+
+### Question: should we store a content hash too?
+
+We're storing a git tree hash in the index, but if we ever want to store package
+contents in the future, maybe we should also store a hash of the git tree
+contents? This would allow verification of package tarballs, without needing the
+whole git repo.
+
+## Integration point: package maps
+
+Other tools (like build systems) may need to (1) consume the dependency resolution
+that we produce, or (2) plug in their own dependencies to the nickel interpreter.
+For (1), they can consume our lock-file, which will be in JSON. This means its
+format needs to be stable.
+
+For (2), we divide the package management implementation into two parts: the `plate`
+command has all the logic for consuming the manifest, fetching packages, and so on.
+The other part is to teach the nickel interpreter about "package maps," which is
+just a map associating a filesystem path to each pair
+`(package filesystem path, package-local name)`. When the nickel interpreter
+is running a file that came from the package at path `/home/jneem/foo`, and that
+file contains `import bar`, the nickel interpreter looks up `(/home/jneem/foo, bar)`
+in the package map to figure out where to import from.
```

Comment:
There are two big technical missing bits in your RFC:
- How does the `nickel` command line understand packages? This is your package-map bit, but extended beyond the interpreter: how do we provide said package map to the interpreter to begin with?
- Where/how are the packages stored on disk (I think this is easier than for most programming languages since we store the packages verbatim, so we don't have this thing where the same version of the same package is actually two different artefact because they were built with two different compilers).

Reply:
- I've added some more details to the section on the CLI interface
- I added an extra section at the end on how we store packages